### PR TITLE
FISH-385 Multitenant OpenID Connect

### PR DESCRIPTION
**Pull request by OndroMih**

### CODE REVIEW

File: docs/modules/ROOT/pages/documentation/payara-server/public-api/openid-connect-support.adoc [hunk 1]
```diff
@@ -251,8 +250,96 @@ public class SecurityBean {
 }
 ----
 
+IMPORTANT: By default, the EL expressions are evaluated only once after the  application is loaded and the evaluated values are then remembered until the application is reloaded, for performance reasons. This means that although the configuration can be evaluated dynamically the first time it's needed, it's not possible to change the configuration later on. If you need to dynamically modify the configuration during the lifetime of the application, follow the next section about multitenancy support.
+
+[[multitenancy]]
+== Multitenancy Support (Session-scoped Configuration)
+
+_Since 5.2021.5_
```
Comment (fturizo) [inline docs/modules/ROOT/pages/documentation/payara-server/public-api/openid-connect-support.adoc:258]: According to the new style guide, we have decided to stop using this disclaimer

File: docs/modules/ROOT/pages/documentation/payara-server/public-api/openid-connect-support.adoc [hunk 2]
```diff
@@ -251,8 +250,96 @@ public class SecurityBean {
...
+
+By default, the same configuration of the OpenID connector is applied for the whole application, for all authentication attempts. This is for performance reasons. The OpenID connector also supports reevaluating the configuration for each user session, before each authentication attempt. This is useful in a multitenant scenario to define a different configuration for each tenant. It's also useful if the user should be able to select which provider they want to use to authenticate.
```
Comment (fturizo) [inline docs/modules/ROOT/pages/documentation/payara-server/public-api/openid-connect-support.adoc:260]: re-evaluating

File: docs/modules/ROOT/pages/documentation/payara-server/public-api/openid-connect-support.adoc [hunk 3]
```diff
@@ -251,8 +250,96 @@ public class SecurityBean {
...
+
+To enable reevaluating the configuration for each user session, set the MicroProfile Configuration property `payara.security.openid.sessionScopedConfiguration` to `true`. To specify it directly in the application, you can place it in the https://download.eclipse.org/microprofile/microprofile-config-1.4/microprofile-config-spec.html#default_configsources[microprofile-config.properties] file in the `META-INF` directory on the classpath (in a WAR application it could be in `WEB-INF/classes/META-INF`).
```
Comment (fturizo) [inline docs/modules/ROOT/pages/documentation/payara-server/public-api/openid-connect-support.adoc:262]: Should be: "To enable the re-evaluation of the configuration per user session, ..."

File: docs/modules/ROOT/pages/documentation/payara-server/public-api/openid-connect-support.adoc [hunk 4]
```diff
@@ -251,8 +250,96 @@ public class SecurityBean {
...
+
+With this enabled, it's possible to use EL expressions to dynamically adjust the configuration before each authantication attempt, e.g. based on any information in the incoming HTTP request. The information about the HTTP request can be retrieved from a `HttpServletRequest` object injected using `@Inject`.
+
+NOTE: It's not possible to use a different configuration for just a subset of secured resources. Once a user is authenticated, the authentication information is saved in the HTTP session. All secured resources will be accessed using the same user, having the same roles, until the user logs out.
+
+=== Example multitenant authentication
+
+In this example, we'll:
+
+* enable session-scoped OpenID Connect configuration
```
Comment (fturizo) [inline docs/modules/ROOT/pages/documentation/payara-server/public-api/openid-connect-support.adoc:272]: Capitalize the sentences in the list

File: docs/modules/ROOT/pages/documentation/payara-server/public-api/openid-connect-support.adoc [hunk 5]
```diff
@@ -251,8 +250,96 @@ public class SecurityBean {
...
+[source,xml]
+----
+payara.security.openid.tenant1.providerURI=<TENANT1_OPENID_PROVIDER_URI>
+payara.security.openid.tenant2.providerURI=<TENANT2_OPENID_PROVIDER_URI>
+payara.security.openid.sessionScopedConfiguration=true
+----
+
+This will provide configuration for `tenant1` and `tenant2` tenants. For each additional tenant, add a new line for its `providerURI`.
+
+Create `OpenidConfigBean` class with the `tokenEndpointURL` method. This class will be a CDI bean that injects `HttpServletRequest` to get information about which tenant to use. It will also inject `Config` to retrieve the configuration about each tenant from the `microprofile-config.properties` file:
```
Comment (fturizo) [inline docs/modules/ROOT/pages/documentation/payara-server/public-api/openid-connect-support.adoc:291]: Create an `OpenIDConfigBean` class

File: docs/modules/ROOT/pages/documentation/payara-server/public-api/openid-connect-support.adoc [hunk 6]
```diff
@@ -251,8 +250,96 @@ public class SecurityBean {
...
+)
+public class SecurityBean {
+}
+----
+
+
 [[logout]]
-=== Logout functionality
+== Logout Functionality
 _Since 5.2020.3_
```
Comment (fturizo) [inline docs/modules/ROOT/pages/documentation/payara-server/public-api/openid-connect-support.adoc:343]: According to the new style guide, we have decided to stop using this disclaimer, so better to remove it now.

**Review by fturizo — CHANGES_REQUESTED** [review]
Looks good, minor changes needed

**Comment by OndroMih** [comment]
FISH-385 is postponed to a later release, do not merge this PR before the upcoming release.

**Pull request merged**